lib/wasi/src/syscalls/mod.rs

```diff
@@ -0,0 +1,218 @@
+use crate::state::WasiState;
+use wasmer_runtime_core::{memory::Memory, vm::Ctx};
+
+#[allow(clippy::mut_from_ref)]
+fn get_wasi_state(ctx: &Ctx) -> &mut WasiState {
+    unsafe { &mut *(ctx.data as *mut WasiState) }
```

Comment:
One idea: we could differentiate between, ModuleState, InstanceState, FunctionState (if we add that).